src/Parsers/ASTLiteral.h

```diff
@@ -33,6 +34,9 @@ class ASTLiteral : public ASTWithAlias
      */
     String unique_column_name;
 
+    /// Hint for data type of literal, that can be set by operator "::".
+    DataTypePtr data_type_hint;
```

Comment:
For better separation of concepts we can do the following:
- parse only type name in parser; don't create an instance of IDataType; place String here;
- parse literals with `::` specification as String literals;
- parse `::` as CAST operator.

This will help to avoid dependency from Parser to DataTypes and Columns.

Reply:
PS. We also need to think about implementation of `123.456::Decimal`.

Reply:
It will also allow to implement faster parsing of Arrays.

Reply:
I've reworked `ParserCastOperator`